File: python/src/bursar/shared/idempotency.py
"""Validation shared by replay-safe Bursar mutations."""

from __future__ import annotations

import hashlib
import re
from typing import Annotated

from pydantic import BeforeValidator, Field

MAX_IDEMPOTENCY_KEY_LENGTH = 255


def require_stable_key(value: object, field: str = "idempotency_key") -> str:
    """Return a valid caller-owned replay key without rewriting it."""
    if not isinstance(value, str) or not value or value != value.strip() or len(value) > MAX_IDEMPOTENCY_KEY_LENGTH:
        raise ValueError(f"{field} must be a trimmed non-empty string of at most 255 characters")
    return value


def scope_stable_key(
    value: object,
    scope: object,
    *identity_parts: object,
    field: str = "idempotency_key",
) -> str:
    """Derive a bounded replay key for one deterministic child mutation."""
    stable_key = require_stable_key(value, field)
    if not isinstance(scope, str) or re.fullmatch(r"[a-z][a-z0-9-]{0,47}", scope) is None:
        raise ValueError("idempotency key scope must match [a-z][a-z0-9-]{0,47}")
    encoded_parts: list[str] = []
    for part in identity_parts:
        if not isinstance(part, str) or not part or part != part.strip():
            raise ValueError("idempotency key identity parts must be trimmed non-empty strings")
        encoded_parts.append(f"{len(part)}#{part}")
    candidate = ":".join((stable_key, scope, *encoded_parts))
    if len(candidate) <= MAX_IDEMPOTENCY_KEY_LENGTH:
        return candidate
    digest = hashlib.sha256(candidate.encode()).hexdigest()
    return f"bursar:{scope}:{digest}"


StableKey = Annotated[
    str,
    BeforeValidator(require_stable_key),
    Field(strict=True),
]

__all__ = [
    "MAX_IDEMPOTENCY_KEY_LENGTH",
    "StableKey",
    "require_stable_key",
    "scope_stable_key",
]

File: javascript/src/shared/idempotency.ts
import { createHash } from "node:crypto";
import { z } from "zod";

const MAX_IDEMPOTENCY_KEY_LENGTH = 255;
const INTERNAL_SCOPE_RE = /^[a-z][a-z0-9-]{0,47}$/u;
type StableKeyInput = string | null | undefined;

/** Validate a caller-owned replay key without silently rewriting it. */
const stableKeySchema = z
  .string()
  .min(1)
  .refine((value) => value === value.trim())
  .refine((value) => Array.from(value).length <= MAX_IDEMPOTENCY_KEY_LENGTH);

export function requireStableKey(value: StableKeyInput, field = "idempotencyKey"): string {
  const parsed = stableKeySchema.safeParse(value);
  if (!parsed.success) {
    throw new TypeError(`${field} must be a trimmed non-empty string of at most 255 characters`);
  }
  return parsed.data;
}

/**
 * Namespace a stable key without truncating it or creating prefix collisions.
 * Dynamic identifiers participate in the candidate hash, while the bounded
 * internal scope keeps the fallback readable and within provider limits.
 */
export function scopedStableKey(
  value: StableKeyInput,
  scope: string,
  components: readonly string[] = [],
): string {
  const key = requireStableKey(value);
  const parsedScope = z.string().regex(INTERNAL_SCOPE_RE).safeParse(scope);
  if (!parsedScope.success) {
    throw new TypeError("scope must be a short lowercase internal label");
  }
  const parsedComponents = z
    .array(
      z
        .string()
        .min(1)
        .refine((component) => component === component.trim()),
    )
    .safeParse(components);
  if (!parsedComponents.success) {
    throw new TypeError("scope components must be non-empty trimmed strings");
  }
  const encodedComponents = parsedComponents.data.map(
    (component) => `${Array.from(component).length}#${component}`,
  );
  const candidate = [key, parsedScope.data, ...encodedComponents].join(":");
  if (Array.from(candidate).length <= MAX_IDEMPOTENCY_KEY_LENGTH) return candidate;
  const digest = createHash("sha256").update(candidate, "utf8").digest("hex");
  return `bursar:${parsedScope.data}:${digest}`;
}
